base_layer/core/src/mempool/utx_pool/utx_pool.rs

```diff
@@ -0,0 +1,316 @@
+    };
+
+    // Create an unconfirmed transaction for testing with a valid fee, unique access_sig and random inputs, the
+    // transaction is only partially constructed
+    fn create_test_tx(fee: MicroTari) -> Transaction {
+        let mut rng = rand::OsRng::new().unwrap();
+
+        let tx_meta = TransactionMetadata {
+            fee: fee.clone(),
+            lock_height: 0,
+        };
+        let key = PrivateKey::random(&mut rng);
+        let r = PrivateKey::random(&mut rng);
+        let e = build_challenge(&PublicKey::from_secret_key(&r), &tx_meta);
+        let s = Signature::sign(key.clone(), r, &e).unwrap();
+        let excess = COMMITMENT_FACTORY.commit_value(&key, 0);
+        let kernel = KernelBuilder::new()
+            .with_fee(fee)
+            .with_lock_height(0)
+            .with_excess(&excess)
+            .with_signature(&s)
+            .build()
+            .unwrap();
+
+        let input1 = TransactionInput {
+            commitment: COMMITMENT_FACTORY.commit(&PrivateKey::random(&mut rng), &MicroTari(10).into()),
+            features: OutputFeatures::default(),
+        };
+        let input2 = TransactionInput {
+            commitment: COMMITMENT_FACTORY.commit(&PrivateKey::random(&mut rng), &MicroTari(50).into()),
+            features: OutputFeatures::default(),
+        };
+
+        let mut body = AggregateBody::empty();
+        body.kernels.push(kernel);
+        body.inputs.push(input1);
+        body.inputs.push(input2);
+
+        Transaction {
+            offset: PrivateKey::random(&mut rng),
+            body,
+        }
+    }
+
+    // Create a partially constructed block for testing
+    fn create_test_block(transactions: Vec<Transaction>) -> Block {
+        let mut body = AggregateBody::empty();
+        transactions.iter().for_each(|tx| {
+            body.kernels.push(tx.body.kernels[0].clone());
+            body.inputs.append(&mut tx.body.inputs.clone());
+        });
+
+        Block {
+            header: BlockHeader::new(0),
+            body,
+        }
+    }
+
+    #[test]
+    fn test_insert_and_retrieve() {
+        let utx1 = create_test_tx(MicroTari(500));
+        let utx2 = create_test_tx(MicroTari(100));
+        let utx3 = create_test_tx(MicroTari(1000));
+        let utx4 = create_test_tx(MicroTari(200));
+        let utx5 = create_test_tx(MicroTari(500));
+
+        let mut utx_pool = UTxPool::new(UTxPoolConfig { storage_capacity: 4 });
+        utx_pool
+            .insert(vec![
+                utx1.clone(),
+                utx2.clone(),
+                utx3.clone(),
+                utx4.clone(),
+                utx5.clone(),
+            ])
+            .unwrap();
+        // Check that lowest priority utx was removed to make room for new incoming transaction
+        assert!(utx_pool.contains_key(&utx1.body.kernels[0].excess_sig).unwrap());
+        assert!(!utx_pool.contains_key(&utx2.body.kernels[0].excess_sig).unwrap());
```

Comment:
nit: I personally find it hard to pick up the second `!`. and prefer `assert_eq!(..., false)` for readability, even though it's tehnically more verbose.

Reply:
I also like to use `assert_eq!(..., Ok(()))` or `assert_eq!(.., Ok(false))` so that you get the diff comparison if it's an Err

Reply:
Changed to `assert_eq!(..., false)`

Reply:
👍